target_chains/solana/programs/pyth-price-publisher/src/instruction.rs

```diff
@@ -0,0 +1,133 @@
+const BUFFER_SEED: &str = "BUFFER";
+
+/// Max number of prices allowed to be stored in a single publisher account.
+const MAX_NUM_PRICES: usize = 128;
+
+#[repr(u8)]
+pub enum Instruction {
+    // key[0] payer     [signer writable]
+    // key[1] vault     [writable]
+    // key[2] system    []
+    Initialize,
+    // key[0] publisher [signer writable]
+    // key[1] buffer    [writable]
+    SubmitPrices,
+    // key[0] publisher []
+    // key[1] vault     [writable]
+    // key[2] buffer    [writable]
+    // key[3] system    []
+    InitializePublisher,
+}
+
+impl Instruction {
+    pub fn parse(input: &[u8]) -> Result<Instruction, ProgramError> {
+        match input.first() {
+            Some(0) => Ok(Instruction::Initialize),
+            Some(1) => Ok(Instruction::SubmitPrices),
+            Some(2) => Ok(Instruction::InitializePublisher),
+            _ => Err(ProgramError::InvalidInstructionData),
+        }
+    }
+}
+
+fn validate_publisher<'a>(
+    account: Option<&AccountInfo<'a>>,
+    require_signer_writable: bool,
+) -> Result<AccountInfo<'a>, ProgramError> {
+    let publisher = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    if require_signer_writable {
+        ensure!(
+            ProgramError::InvalidInstructionData,
+            publisher.is_signer,
+            publisher.is_writable
+        );
+    }
+    Ok(publisher)
+}
+
+fn validate_system<'a>(account: Option<&AccountInfo<'a>>) -> Result<AccountInfo<'a>, ProgramError> {
+    let system = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    ensure!(
+        ProgramError::InvalidInstructionData,
+        system_program::check_id(system.key)
+    );
+    Ok(system)
+}
+
+fn validate_payer<'a>(account: Option<&AccountInfo<'a>>) -> Result<AccountInfo<'a>, ProgramError> {
+    let payer = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    ensure!(
+        ProgramError::InvalidInstructionData,
+        payer.is_signer,
+        payer.is_writable
+    );
+    Ok(payer)
+}
+
+fn pubkey_eq(a: &Pubkey, b: &Pubkey) -> bool {
+    sol_memcmp(a.as_ref(), b.as_ref(), 32) == 0
+}
+
+fn validate_vault<'a>(
+    account: Option<&AccountInfo<'a>>,
+    program_id: &Pubkey,
+) -> Result<(AccountInfo<'a>, u8), ProgramError> {
+    let vault = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    let vault_pda = Pubkey::find_program_address(&[VAULT_SEED.as_bytes()], program_id);
+    ensure!(
+        ProgramError::InvalidInstructionData,
+        pubkey_eq(&vault.key, &vault_pda.0),
+        vault.is_writable
+    );
+    Ok((vault, vault_pda.1))
+}
+
+fn validate_buffer<'a>(
+    account: Option<&AccountInfo<'a>>,
+    publisher: &Pubkey,
+    program_id: &Pubkey,
+) -> Result<(AccountInfo<'a>, u8), ProgramError> {
+    let buffer = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    let buffer_pda =
```

Comment:
same comment as above, i recommend either get bump u8 this as input instruction or put it in the account. then you can use `create_program_address` to only perform a single check.

Reply:
Replaced with `create_program_address` and added bumps to the instruction arguments.